<!DOCTYPE html>
<html>
  <head>
    <title>{{ reportName }}&mdash;Pathway Analysis Report</title>
    <style>
      @import url("https://rsms.me/inter/inter.css");
      html {
        font-family: "Inter", sans-serif;
      }

      @supports (font-variation-settings: normal) {
        html {
          font-family: "Inter var", sans-serif;
        }
      }

      body {
        color: #5e6c84;
        letter-spacing: 0.012em;

        font-size: 12px;
        line-height: 16px;
      }

      h1 {
        color: #000;
        margin: 0;
        margin-top: 20px;
        margin-bottom: 20px;
        font-weight: 400;

        font-size: 24px;
        line-height: 32px;
      }

      h2 {
        color: #101426;
        margin: 0;
        margin-bottom: 8px;
        font-weight: 400;

        font-size: 16px;
        line-height: 20px;
      }

      h3 {
        color: #101426;
        margin: 0;
        margin-bottom: 8px;
        font-weight: 400;

        line-height: 15px;
      }

      p {
        margin: 0;
        margin-bottom: 12px;
      }

      .f {
        display: flex;
      }

      .c {
        flex-direction: column;
      }

      .m {
        max-width: 800px;
        margin: auto;
      }

      .s {
        margin-bottom: 60px;
      }

      .store {
        display: none;
      }
    </style>
  </head>
  <!-- TODO(aidan): add fallbacks -->
  <body>
    <div class="f c m">
      <h1>{{ reportName }}</h1>
      Gene set enrichment analysis (GSEA) was performed on three databases:
      Gene Ontology (GO), Kyoto Encyclopedia of Genes and Genomes (KEGG), and 
      the Molecular Signatures Database (MSigDB).
      <br /><br /><br />
      <div class="f c s">
        <h2>Gene Ontology</h2>
        <div>
          The <a href='http://geneontology.org/'>Gene Ontology</a> resource provides a computational representation of our current scientific knowledge about the functions of genes (or, more properly, the protein and non-coding RNA molecules produced by genes) from many different organisms, from humans to bacteria. Our
          analyses used data drawn from this <a href="http://geneontology.org/">database</a>.
        </div>
        <img src="./Gene Ontology/Dot Plot.png" />
        <img src="./Gene Ontology/Ridge Plot.png" />
      </div>
      <div class="f c s">
        <h2>KEGG</h2>
        <div>
          KEGG is a database resource for understanding high-level functions and utilities of the biological system, such as the cell, the organism and the ecosystem, from molecular-level information, especially large-scale molecular datasets generated by genome sequencing and other high-throughput experimental technologies. Our 
          analyses used data drawn from the
          <a href='https://www.genome.jp/kegg/pathway.html'>KEGG Pathway Database</a>
        </div>
        <img src="./KEGG/Dot Plot.png" />
        <img src="./KEGG/Ridge Plot.png" />
      </div>
      <div class="f c s">
        <h2>MSig</h2>
        <div>
          The Molecular Signatures Database (MSigDB) is a collection of annotated gene sets for use with GSEA software. Our analyses used data drawn from this 
          <a href='http://www.gsea-msigdb.org/gsea/msigdb'>database</a>.
        </div>
        <img src="./MSig/Dot Plot.png" />
        <img src="./MSig/Ridge Plot.png" />
      </div>
      <div class="store">
        {% for data in pathviews %}
          <img id="pathview-{{ data.id }}" src="{{ data.path }}" />
        {% endfor %}
      </div>
      <div id="react-entrypoint"></div>
    </div>
  </body>
</html>
